<center>
<h1> TP-Projet d'optimisation numérique </h1>
<h1> Algorithme de Newton </h1>
</center>

## Implémentation 
 
1. Coder l’algorithme de Newton dans le fichier `src/newton.jl` en respectant la spécification donnée dans ce même fichier ;
2. Exécuter les tests ci-dessous et vérifier qu'ils passent.

Pour les tests, nous avons défini les fonctions suivantes $f_1 \colon \mathbb{R}^3 \to \mathbb{R}$
et $f_2 \colon \mathbb{R}^2 \to \mathbb{R}$.

$$
    f_{1}(x_1, x_2, x_3) = 2 (x_1 +x_2 + x_3 -3)^2 + (x_1-x_2)^2 + (x_2 - x_3)^2
$$
et
$$
    f_{2}(x_1,x_2) = 100(x_2-x_1^2)^2 + (1-x_1)^2.
$$

**Remarque.** On peut retrouver ces fonctions dans le fichier `test/fonctions_de_tests.jl`.

In [1]:
include("../src/newton.jl")         # votre algorithme de Newton
include("../test/tester_newton.jl") # la fonction pour tester votre algorithme de Newton

#
afficher = true # si true, alors affiche les résultats des algorithmes

#
tester_newton(newton, afficher); # tester l'algorithme de Newton

Affichage des résultats des algorithmes : true

-------------------------------------------------------------------------
Résultats de : newton appliqué à f1:
  * x0       = [1, 1, 1]
  * x_sol    = [1, 1, 1]
  * f(x_sol) = 0
  * nb_iters = 0
  * flag     = 0
  * solution = [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : newton appliqué à f1:
  * x0       = [1, 0, 0]
  * x_sol    = [1.0, 1.0, 0.9999999999999999]
  * f(x_sol) = 1.232595164407831e-32
  * nb_iters = 1
  * flag     = 0
  * solution = [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : newton appliqué à f1:
  * x0       = [10.0, 3.0, -2.2]
  * x_sol    = [1.0, 0.9999999999999996, 0.9999999999999987]
  * f(x_sol) = 7.296963373294359e-30
  * nb_iters = 1
  * flag     = 0
  * solution = [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : newton appliqué à f2:
  * x0       = [-1.2, 1.0

In [2]:
include("../src/newton.jl") # votre algorithme de Newton
include("../test/fonctions_de_tests.jl") # pour avoir la fonction d'affichage des résultats

# Fonction f0
# -----------
f0(x)      =  sin(x)
grad_f0(x) =  cos(x) # la gradient de la fonction f0
hess_f0(x) = -sin(x) # la hessienne de la fonction f0
solution   = -pi/2

x0 = solution
x_sol, f_sol, flag, nb_iters = newton(f0, grad_f0, hess_f0, x0)
afficher_resultats("Newton", "f0", x0, x_sol, f_sol, flag, nb_iters, solution)

x0 = -pi/2+0.5
x_sol, f_sol, flag, nb_iters = newton(f0, grad_f0, hess_f0, x0)
afficher_resultats("Newton", "f0", x0, x_sol, f_sol, flag, nb_iters, solution)

x0 = pi/2
x_sol, f_sol, flag, nb_iters = newton(f0, grad_f0, hess_f0, x0)
afficher_resultats("Newton", "f0", x0, x_sol, f_sol, flag, nb_iters, solution)


-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0:
  * x0       = -1.5707963267948966
  * x_sol    = -1.5707963267948966
  * f(x_sol) = -1.0
  * nb_iters = 0
  * flag     = 0
  * solution = -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0:
  * x0       = -1.0707963267948966
  * x_sol    = -1.5707963267949088
  * f(x_sol) = -1.0
  * nb_iters = 3
  * flag     = 0
  * solution = -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0:
  * x0       = 1.5707963267948966
  * x_sol    = 1.5707963267948966
  * f(x_sol) = 1.0
  * nb_iters = 0
  * flag     = 0
  * solution = -1.5707963267948966


## Interprétation 

1. Justifier les résultats obtenus pour l'exemple $f_0$ ci-dessus;
2. Justifier que l’algorithme implémenté converge en une itération pour $f_{1}$;
3. Justifier que l’algorithme puisse ne pas converger pour $f_{2}$ avec certains points initiaux.

**Remarque.** Vous pouvez mettre `affiche=true` dans les tests de l'algorithme de Newton pour
vous aider.


## Réponses
**Question 1**

$$
𝑓_0(x) = sin(x)
$$


- Dans le cas où $x_0 = -\pi/2$:
    - Le critère d'arrêt CN1 est atteint, sans aucune itération, car le point de départ, x0, est égal à la solution, et donc la norme du gradient ∥∇f(x0)∥ est proche de zéro.
- Dans le cas où $x_0 = -\pi/2+0.5$:
    - Après 3 itérations, nous trouvons que la norme du gardient converge vers 0, et donc le critère d'arrêt CN1 est satisfait, ce qui confirme que le point de départ est proche à la solution.
- Dans le cas où $x_0 = \pi/2$:
    - Le critère d'arrêt CN1 est satisfait, sans faire d'itération, car le point départ est l'opposé de la solution et donc : ∥∇f(x0)∥ = ∥∇f(pi/2)∥ = ∥∇f(-pi/2)∥ ≈ 0

**Question 2**

Nous avons la fonction suivante :
$$
    f_{1}(x_1, x_2, x_3) = 2 (x_1 +x_2 + x_3 -3)^2 + (x_1-x_2)^2 + (x_2 - x_3)^2
$$
Calculons son gradient $\nabla f_{1}$ et sa hessienne $\nabla^2 f_{1}$ de $f_{1}(x_1, x_2, x_3)$ :

$$
\nabla f_{1} = \begin{bmatrix}
4 (x_1 + x_2 + x_3 - 3) + 2 (x_1 - x_2) \\
4 (x_1 + x_2 + x_3 - 3) - 2 (x_1 - x_2) + 2 (x_2 - x_3) \\
4 (x_1 + x_2 + x_3 - 3) - 2 (x_2 - x_3)
\end{bmatrix}
$$

$$
\nabla^2 f_{1} = \begin{bmatrix}
6&2&4 \\
2&8&2 \\
4&2&6
\end{bmatrix}
$$
La fonction $f_{1}(x_1, x_2, x_3)$ est une fonction quadratique, son développement de Taylor jusqu'à l'ordre 2 est exact:
$$
    f_{1}(x) ≈ f_{1}(a) + \nabla f_{1}(a)^T (x-a) + 1/2 (x-a)^T \nabla^2 f_{1}(a) (x-a)
$$
Et puisque la Hessienne $\nabla^2 f_{1}$ est constante, définie positive et inversible, la fonction $f_{1}(x)$ est convexe. En conséquence, le minimum de $f_{1}(x)$ est atteint lorsque $\nabla f_{1} = 0$. La solution exacte est donc obtenue en une seule itération de l'algorithme de Newton, ce qui résulte en une convergence rapide vers la solution optimale.

**Question 3**

$$
    f_{2}(x_1,x_2) = 100(x_2-x_1^2)^2 + (1-x_1)^2.
$$
Pour les deux premiers cas, $x_{0} = [-1.2; 1]$ et $x_{0} = [10; 0]$, l'algorithme converge vers la solution $x_{sol} = [1,1]$ en un nombre raisonnable d'itérations. Mais pour le cas où le point initial $x_{0} = [0; 1/200 + 1/10^{12}]$, l'algorithme ne converge pas, indiqué par le $flag = 3$, car il atteint nombre maximal d'itérations sans satisfaire les critères d'arrêt.

Calculons la hessienne de la fonction $f_{2}$ :
$$
\nabla^2 f_{1} = \begin{bmatrix}
-400(x_{2}-3x_{1}^2)+2&-400x_{1}\\
-400x_{1}&200
\end{bmatrix}
$$
Avec le point $x_{0} = [0; 1/200 + 1/10^{12}]$, la hessienne devient :
$$
\nabla^2 f_{1} = \begin{bmatrix}
2&0\\
0&200
\end{bmatrix}
$$
Dans ce cas, la hessienne devient essentiellemnt diagonale avec des valeurs propres $2$ et $200$. La valeur propre 200 est particulièrement grande par rapport à 2, ce qui indique que la hessienne est mal conditionnée, et donc les directions de descente calculées par l'algorithme de Newton peut être imprécises.